dagql/types.go

```diff
@@ -864,6 +865,132 @@ type EnumValue interface {
 	~string
 }
 
+type Enum interface {
+	Typed
+	ListValues() ast.EnumValueList
+}
+
+type DynamicEnumValue struct {
+	parent *DynamicEnumValues
+	value  string
+}
+
+func NewDynamicEnumValue(parent *DynamicEnumValues, value string) *DynamicEnumValue {
+	return &DynamicEnumValue{parent, value}
+}
+
+func (e *DynamicEnumValue) Type() *ast.Type {
+	return e.parent.Type()
+}
+
+func (e *DynamicEnumValue) TypeDescription() string {
+	if isType, ok := any(e.parent).(Descriptive); ok {
+		return isType.TypeDescription()
+	}
+
+	return ""
+}
+
+func (e *DynamicEnumValue) Decoder() InputDecoder {
+	return e.parent
+}
+
+func (e *DynamicEnumValue) ToLiteral() call.Literal {
+	return e.parent.Literal(e.value)
+}
+
+func (e *DynamicEnumValue) Value() string {
+	return e.value
+}
+
+type DynamicEnumValues struct {
+	t      Enum
+	values []string
+}
+
+func NewDynamicEnum(t Enum) *DynamicEnumValues {
+	enumValues := t.ListValues()
+
+	values := make([]string, len(enumValues))
+	for _, v := range enumValues {
+		values = append(values, v.Name)
+	}
+
+	return &DynamicEnumValues{t: t, values: values}
+}
```

Comment:
Something about this still doesn't smell quite right - we shouldn't be taking an `ast` value as input in any way here - `dagql` should inform what `ast` is output, not used to inform what is input.

I think the `ListValues` in the `Enum` interface is kinda the source of this. Why do we need to have the `Enum` interface at all?

IMO, the way we should have this is that `NewDynamicEnum` should take a list of `[]DynamicEnumValue` - so it's essentially the "at-runtime" version of `NewEnum` that doesn't use generics.

Reply:
I agree with you, however this would require to loop on all possible values to construct them, which make it quite ugly in the code

Reply:
It'll surely just add a few lines no? Where specifically will it make it ugly?

For example, in `core/typedef.go` it's still very simple - we don't give any possible values, since those aren't known at that point.

Also, in general, go prefers a more explicit style - it's okay to be a bit long-winded, especially in a case like this, where the "neater" approach is more difficult to understand and has unclear dependencies.

Reply:
> Also, in general, go prefers a more explicit style - it's okay to be a bit long-winded, especially in a case like this, where the "neater" approach is more difficult to understand and has unclear dependencies.

Alright, I'll do the changes

Reply:
Removed this, and essentially moved `DynamicEnum` into `core` - it's not a base "dagql" type like everything else, it's specifically much more like an object, which is defined in `core` specifically for modules to expose.